### ISSUE TITLE
During social sign on request with spring security if a user revokes permission if user a user cancels/revokes permission but authentication still succeeds

### ISSUE THREAD

**Issue by airduster** [pac4j/spring-security-pac4j#28]
In ClientAuthenticationFilter.class#attemptAuthentication(...) and spring security when a user revokes permission the returned response should invoke getFailureHandler().onAuthenticationFailure(...)  instead of a getSuccessHandler().onAuthenticationSuccess(...). 

if you have a .defaultSuccessUrl or .successHandler(...) defined in your spring security config like so

**Sample java spring security config.**
 ```           .and()
            	.addFilterBefore(clientAuthenicationFilter(), AnonymousAuthenticationFilter.class)
            	.formLogin()
                	.permitAll()
                	.loginPage		   ( springSecurityProperties().getLoginPage() )
                	.loginProcessingUrl( springSecurityProperties().getLoginProcessingUrl() )
                	**.defaultSuccessUrl.** ( springSecurityProperties().getDefaultTargetUrl() )
                	.failureUrl		   ( springSecurityProperties().getAuthenticationFailureUrl() )
                	.usernameParameter ( springSecurityProperties().getUsernameParameter() )
                	.passwordParameter ( springSecurityProperties().getPasswordParameter() )
                	//.successHandler(authenticationSuccessHandler) 
```

**The problem is.** 

If the user cancels/revokes permission to sign on with a social login with the above config... the ClientAuthenticationFilter#attemptAuthentication(...) method will invoke getSuccessHandler().onAuthenticationSuccess(...) even though the User rejected every permission scope requested to access their account. Thus, this creates a "false positive" login scenario where the spring security framework thinks the user has logged in successfully but in reality we did not receive any information about the user on the call back url. 

**Sample Log:**

* DEBUG:   org.springframework.security.web.util.matcher.AntPathRequestMatcher - Checking match of request : '/callback'; against '/callback'
* DEBUG:   org.pac4j.springframework.security.web.ClientAuthenticationFilter - Request is to process authentication
* DEBUG:   org.pac4j.oauth.client.BaseOAuthClient - **authentication has been cancelled by user**
* DEBUG:   org.pac4j.springframework.security.web.ClientAuthenticationFilter - **credentials : null**
* DEBUG: org.springframework.security.web.authentication.SavedRequestAware**AuthenticationSuccessHandler** - Using default Url: / 

**Proposed solution**: to invoke getFailureHandler().onAuthenticationFailure(...)  instead of getSuccessHandler().onAuthenticationSuccess(...) In ClientAuthenticationFilter.class#attemptAuthentication(...)

**Comment by leleuj** [comment]
PR: https://github.com/pac4j/spring-security-pac4j/pull/29

**Comment by airduster** [comment]
looks good thanks. I'll try it out when the snapshot becomes available.

**Comment by leleuj** [comment]
I just merged the PR, the new SNAPSHOT will be available in a few minutes. I will close the issue as well. Re-open it if you encounter any problem.

**Comment by airduster** [comment]
Hi Jérôme, 

I've been testing this fix and I discovered another bug (or a maybe misconfiguration on my part). I didn't realize that the ClientAuthenticationFilter.class initialized by default its own 

```
...
private AuthenticationSuccessHandler successHandler = new SavedRequestAwareAuthenticationSuccessHandler();
	private AuthenticationFailureHandler failureHandler = new SimpleUrlAuthenticationFailureHandler();
...
```

and in order to get pac4j to utilize your own custom handlers you have to explicitly set them on the **ClientAuthenticationFilter.class#setAuthenticationFailureHandler(...)** in the bean config declaration. 

For some reason I thought (incorrectly) that the Success/Failure handlers were defined on the AuthenticationManager and pac4j was "inheriting" them from spring security.

# Problem:

That being said the new bug is: So by default spring security sets a default authenticationFailureUrl value of:  "/login?error=1" and internally sets it on the AuthenticationFailureHandler that it creates. 

see: [/src/main/java/org/springframework/security/config/annotation/web/configurers/AbstractAuthenticationFilterConfigurer.java](https://github.com/spring-projects/spring-security/blob/c8f598778f3d45e0092385d20593ade6f33e138f/config/src/main/java/org/springframework/security/config/annotation/web/configurers/AbstractAuthenticationFilterConfigurer.java) # public final T failureHandler

and 

[/config/src/main/java/org/springframework/security/config/http/FormLoginBeanDefinitionParser.java](https://github.com/spring-projects/spring-security/blob/205ef42cfbac3cb5d10c6d1e526fed403c51142e/config/src/main/java/org/springframework/security/config/http/FormLoginBeanDefinitionParser.java) (cntrl + f DEF_FORM_LOGIN_AUTHENTICATION_FAILURE_URL)

So because ClientAuthenticationFilter.class is creating it's own default SimpleUrlAuthenticationFailureHandler() without passing in a default url value in the constructor. if we didn't explicitly override this object... the default url will be null and throw a nasty NullPointerException when triggered at runtime. This was triggered when testing for user cancelation of permission during facebook login. 

```
java.lang.NullPointerException
	at org.springframework.security.web.authentication.SimpleUrlAuthenticationFailureHandler.onAuthenticationFailure(SimpleUrlAuthenticationFailureHandler.java:63)
	at org.pac4j.springframework.security.web.ClientAuthenticationFilter.attemptAuthentication(ClientAuthenticationFilter.java:99)
```

# Proposed fix: 

either explicitly set the default value to match Spring security's default "/login?error=1" or during initialization put in an Assert.notNull(...) check to make sure the developer sets their own value for SimpleUrlAuthenticationFailureHandler#defaultFailureUrl to avoid the nasty NullPointerException

**Comment by leleuj** [comment]
Indeed, I messed up when testing. I get also a `NullPointerException` but I don't share the same analysis. For me, it's due to the fact we have a `null` exception when calling the authentication failure handler.

I just committed a fix. Can you re-test? Thanks

**Comment by airduster** [comment]
Yes the exception is being handled now I see the 

**HTTP Status 401 - Authentication Failed: No credentials  page**

But... (for me) the root cause of the exception is still because the defaultFailureUrl attribute on SimpleUrlAuthenticationFailureHandler.class was not initialized and when I step into the code is it still null.

And I believe it is because ClientAuthenticationFilter is initialization is an empty constructor arg 

```
new SimpleUrlAuthenticationFailureHandler();
```
instead of 

```
SimpleUrlAuthenticationFailureHandler( SPRING_SECURITY_DEFAULT_FAILURE_URL ) ;
```
Is the default value set for your security configuration?

(if yes)
Then there must be something wrong with my config. Do you mind posting your config? 

(if no)
So I guess the question is should the defaultFailureURL be a required or optional field? Should spring-security-pac4j duplicate the same defaultFailureUrl as spring-security and set the field or should the developer just be aware that it is not set and they have to explicitly inject their own AuthenticationFailureHandler?

**Comment by leleuj** [comment]
Do you still get a `NullPointerException`? I don't think so.

Here is the demo I use for my tests: https://github.com/pac4j/spring-security-pac4j-demo It should be straightforward to launch it on your side and compare the behaviour.

As the `ClientAuthenticationFilter` does not the create the `AuthenticationFailureHandler`, but its Spring Security parent class does, I tend to keep things "as is" to follow possible evolutions of the Spring Security behaviour. But I'm fairly open.

**Comment by airduster** [comment]
Sorry correction the parent class is creating it. 

ClientAuthenticationFilter extends **AbstractAuthenticationProcessingFilter** 

and I can replicate the error with the following config. 

defaultFailureUrl will be null 

```
	@Bean
	public ClientAuthenticationFilter clientAuthenicationFilter() throws Exception {
		
		ClientAuthenticationFilter caf = new ClientAuthenticationFilter(("/callback"));
		caf.setClients(clients());
		caf.setAuthenticationManager( authenticationManager() );
		
		return caf;
	}
```

but if we explicitly set the AuthenticationFailureHandler it Will redirect to explicitly passed in defaultFailureUrl

```
	@Bean
	public ClientAuthenticationFilter clientAuthenicationFilter() throws Exception {
		
		ClientAuthenticationFilter caf = new ClientAuthenticationFilter(("/callback"));
		caf.setClients(clients());
		caf.setAuthenticationManager( authenticationManager() );
		caf.setAuthenticationFailureHandler( new SimpleUrlAuthenticationFailureHandler("/login?error=1"));
		
		return caf;
	}
```

but I was under the same impression as you that handlers should be inherited and if I put a breakpoint in SimpleUrlAuthenticationFailureHandler() it hits ( 2 ) times but i'm not sure where they are getting created. Which leads me to suspect something is wrong with my configuration. 

Yea I wish there was a java config spring security example with pac4j. I tried to look everywhere for one.

**Comment by airduster** [comment]
Spring social project also extends AbstractAuthenticationProcessingFilter and they explicitly pass in their own default handler urls. 

see: 

[SocialAuthenticationFilter.java](https://github.com/spring-projects/spring-social/blob/master/spring-social-security/src/main/java/org/springframework/social/security/SocialAuthenticationFilter.java)

**Comment by leleuj** [comment]
Yes, I should convert the spring-security-pac4j-demo with Java configuration: any help will be appreciated on this.

My feeling is that it works almost as usual in the pac4j ecosystem: no credentials -> 401 error page. So I prefer to stick to that.
In your case, you can define the authentication failure url with the appropriate failure url, right? And it works without causing any bug or stacktrace, doesn't it?
As you are the only one to know your failure url, I think it's more on your side to setup this part.

So I tend to leave things that way and do the release next week. If don't have any objection...

**Comment by airduster** [comment]
Jérôme, 

Apologies for the delayed response I have been on vacation + other project issues that were blocking this one. 

Yes, I can expicitly define the authentication failure handler like so: 

```
ClientAuthenticationFilter caf = new ClientAuthenticationFilter(("/callback"));
		caf.setClients(clients());
		caf.setAuthenticationManager( authenticationManager() );
		caf.setAuthenticationFailureHandler( new SimpleUrlAuthenticationFailureHandler("/login?error=1"));
```

What's strange though (from the limited testing I did) is I don't have to define a successHandler... it redirects me to the homePage. Not sure if this is a default value or if it is using my defined successHandler (which is also the home page '/')

If you have a framework "ethos" and want to be consistent with it I have no problems with that. I would just say in the documentation this should be noted that the failure handler must be explicitly defined. 

Also, I see you've updated the documentation with a Java config, awesome. 

I still need to do further testing (i've been neck deep in other issues). When I finish i'll follow up with you and let you know my findings. 

Thanks, 

S.J

**Comment by leleuj** [comment]
OK. Keep us posted. I'll do the release very soon. I updated the doc: https://github.com/pac4j/spring-security-pac4j/commit/06d7d8cd4d9a5c3a04e4ea3388c678f43556d24e

**Comment by airduster** [comment]
Confirmed, The current implementation you have is correct. 

So according to the javadoc for: 
[org.springframework.security.web.authentication.](https://github.com/spring-projects/spring-security/blob/master/web/src/main/java/org/springframework/security/web/authentication/AbstractAuthenticationProcessingFilter.java)**AbstractAuthenticationProcessingFilter**

The **default** behavior of:

 successHandler --> '/'
 errorHandler --> HTTP 401 

is the correct **default** behavior. 

**Authentication Success**
```
If authentication is successful, the resulting Authentication object will be placed into the SecurityContext 
for the current thread, which is guaranteed to have already been created by an earlier filter.  The configured 
AuthenticationSuccessHandler will then be called to take the redirect to the appropriate destination after a successful login. 
The default behaviour is implemented in a SavedRequestAwareAuthenticationSuccessHandler which will make use of any 
DefaultSavedRequest set by the ExceptionTranslationFilter and redirect the user to the URL contained therein. 
Otherwise it will redirect to the webapp root "/". You can customize this behaviour by injecting a differently 
configured instance of this class, or by using a different implementation. 

See the successfulAuthentication(HttpServletRequest, HttpServletResponse, FilterChain, Authentication) method 
for more information. 
```

**Authentication Failure**
```
If authentication fails, it will delegate to the configured AuthenticationFailureHandler to allow the failure information 
to be conveyed to the client. The default implementation is SimpleUrlAuthenticationFailureHandler , 
which sends a 401 error code to the client. It may also be configured with a failure URL as an alternative. 
Again you can inject whatever behaviour you require here. 
```

**Comment by leleuj** [comment]
Good. Thanks for checking.

**Comment by leleuj** [comment]
I released spring-security-pac4j v1.4.1
